### PR TITLE
search ETag case insensitively

### PR DESCRIPTION
**Pull request by reshke**
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

Some cloud providers form ETAG in a different format, for example Yandex Cloud Object Storage sends it in format
`Etag: "......"`

according to the RFC (https://www.w3.org/Protocols/rfc2616/rfc2616.html)
ETag can be in any case. So we should parse header case insensitively.

also added NextMarker processing logic

### CODE REVIEW

File: gpcontrib/gpcloud/src/s3interface.cpp [hunk 1]
```diff
@@ -470,8 +470,11 @@ string S3InterfaceService::uploadPartOfData(S3VectorUInt8 &data, const S3Url &s3
     Response resp = this->putResponseWithRetries(urlWithQuery.str(), headers, data);
     if (resp.getStatus() == RESPONSE_OK) {
         string headers(resp.getRawHeaders().begin(), resp.getRawHeaders().end());
+        string toSearch = "etag: ";
 
-        uint64_t etagStartPos = headers.find("ETag: ") + 6;
+        uint64_t etagStartPos = std::search(headers.begin(), headers.end(), toSearch.begin(), toSearch.end(), [](char ch1, char ch2){
```
Comment (x4m) [inline gpcontrib/gpcloud/src/s3interface.cpp:475]: Let's check -1 result of std::search()?

File: gpcontrib/gpcloud/src/s3interface.cpp [hunk 2]
```diff
@@ -470,8 +470,11 @@ string S3InterfaceService::uploadPartOfData(S3VectorUInt8 &data, const S3Url &s3
...
+             return std::tolower(ch1) == std::tolower(ch2);
+        }) - headers.begin() + 6;
```
Comment (x4m) [inline gpcontrib/gpcloud/src/s3interface.cpp:477]: maybe sizeof(toSearch) would be slightly better than 6?

**Review by x4m — CHANGES_REQUESTED** [review]
Hi!
This PR fixes obvious incompatibility with RFC.
Meanwhile, maybe, let's make nearby code a little better?

File: gpcontrib/gpcloud/src/s3interface.cpp [hunk 3]
```diff
@@ -262,7 +271,8 @@ bool S3InterfaceService::parseBucketXML(ListBucketResult *result, xmlParserCtxtP
         cur = cur->next;
     }
 
-    marker = (is_truncated && key) ? key : "";
+    if (is_truncated && next_marker) marker = next_marker;
```
Comment (x4m) [inline gpcontrib/gpcloud/src/s3interface.cpp:274]: Is this formatting per project style? I could not find similar formatting of ```if-else``` statement in nearby code.

File: gpcontrib/gpcloud/src/s3interface.cpp [hunk 4]
```diff
@@ -470,8 +484,17 @@ string S3InterfaceService::uploadPartOfData(S3VectorUInt8 &data, const S3Url &s3
...
     if (resp.getStatus() == RESPONSE_OK) {
         string headers(resp.getRawHeaders().begin(), resp.getRawHeaders().end());
+        string toSearch = "etag: ";
+
+        auto res = std::search(headers.begin(), headers.end(), toSearch.begin(), toSearch.end(), [](char ch1, char ch2){
+             return std::tolower(ch1) == std::tolower(ch2);
+        });
+
+        if (res == headers.end()) {
+            S3_DIE(S3RuntimeError, "responce does not contain etag in header");
```
Comment (x4m) [inline gpcontrib/gpcloud/src/s3interface.cpp:494]: responce -> response

File: gpcontrib/gpcloud/src/s3url.cpp [hunk 5]
```diff
@@ -151,4 +151,4 @@ string S3Url::getExtension() const {
         return "";
     }
     return filename.substr(pos);
-}
\ No newline at end of file
+}
```
Comment (x4m) [inline gpcontrib/gpcloud/src/s3url.cpp:154]: This change is correct, but unrelated.

File: gpcontrib/gpcloud/src/s3interface.cpp [hunk 6]
```diff
@@ -470,8 +484,17 @@ string S3InterfaceService::uploadPartOfData(S3VectorUInt8 &data, const S3Url &s3
     Response resp = this->putResponseWithRetries(urlWithQuery.str(), headers, data);
     if (resp.getStatus() == RESPONSE_OK) {
         string headers(resp.getRawHeaders().begin(), resp.getRawHeaders().end());
+        string toSearch = "etag: ";
+
+        auto res = std::search(headers.begin(), headers.end(), toSearch.begin(), toSearch.end(), [](char ch1, char ch2){
+             return std::tolower(ch1) == std::tolower(ch2);
+        });
```
Comment (zhaorui20005) [inline gpcontrib/gpcloud/src/s3interface.cpp:491]: Could "Etag  " string be in other format?

File: gpcontrib/gpcloud/src/s3interface.cpp (hunk 4 above)
Comment (reshke) [reply gpcontrib/gpcloud/src/s3interface.cpp]: thnx, fixed

File: gpcontrib/gpcloud/src/s3interface.cpp (hunk 3 above)
Comment (reshke) [reply gpcontrib/gpcloud/src/s3interface.cpp]: thnx, fixed

**Review by x4m — APPROVED** [review]

**Comment by zhaorui20005** [comment]
After I merged this pr on master in my envs, I found it cannot get s3 files correctly.
Take an s3 example, assume we create s3 file like this: 

aws s3 ls us-east-1.s3test.pivotal.io/gpss/ --recursive
2021-09-17 14:55:57          7 gpss/test/photos/2020/photo0
2021-09-17 14:56:06          7 gpss/test/photos/2020/photo1
2021-09-17 14:56:16          7 gpss/test/photos/2020/photo2
2021-09-17 16:14:11          7 gpss/test/photos/2021/photo4
2021-09-17 14:54:08         11 gpss/test/sample.jpg
gpcheckcloud -d "s3://s3-us-east-1.amazonaws.com/us-east-1.s3test.pivotal.io/gpss/test config=/tmp/s3.conf"
photo0
photo1
photo2
photo4
sample.jpg

The above output are what we expect to get.
But after I merged this pr and recompile the gpcheckcloud, it cannot get the expect result, so this pr will make a regression for normal function.

**Comment by adam8157** [comment]
@zhaorui20005 what output do you see with this PR? Do you know the cause?

**Comment by zhaorui20005** [comment]
> @zhaorui20005 what output do you see with this PR? Do you know the cause?

The valuable codes are added between 490-500 in s3interface.cpp which enables the s3 to recognize the **case insensitive** "ETag"in the response, because some servers which support S3 APIs may return "etag" or "Etag" which cause the s3 failed to recognize the "etag" part and throw an exception.

**Comment by adam8157** [comment]
Quote:

    But after I merged this pr and recompile the gpcheckcloud, it cannot get the expect result, so this pr will make a regression for normal function.

You said it can not get the expected result, what output did it get?

**Comment by zhaorui20005** [comment]
> Quote:
> 
> ```
> But after I merged this pr and recompile the gpcheckcloud, it cannot get the expect result, so this pr will make a regression for normal function.
> ```
> 
> You said it can not get the expected result, what output did it get?

Empty result. But we expect it can get the file data.

File: gpcontrib/gpcloud/src/s3interface.cpp [hunk 7]
```diff
@@ -292,13 +305,17 @@ ListBucketResult S3InterfaceService::listBucket(S3Url &s3Url) {
 
         // transfer /bucket/prefix to /bucket/?prefix=prefix because we need to "GET" a real thing
         stringstream querySs;
+
+        querySs << "delimiter" << "=" << UriEncode("/");
```
Comment (adam8157) [inline gpcontrib/gpcloud/src/s3interface.cpp:309]: I guess this is the cause that @zhaorui20005 got empty results.

Will confirm.

**Review by adam8157 — COMMENTED** [review]
Will pair with @zhaorui20005 offline to fix these small issues.

File: gpcontrib/gpcloud/src/s3interface.cpp (hunk 7 above)
Comment (adam8157) [reply gpcontrib/gpcloud/src/s3interface.cpp:309]: Confirmed, removed this logic.

**Comment by adam8157** [comment]
I removed the "delimiter" logic since it has a regression.

Thanks for the contribution.

**Pull request merged**